chiselFrontend/src/main/scala/chisel3/core/UserModule.scala

```diff
@@ -67,9 +67,15 @@ abstract class UserModule(implicit moduleCompileOptions: CompileOptions)
     val firrtlPorts = getModulePorts map {port => Port(port, port.specifiedDirection)}
     _firrtlPorts = Some(firrtlPorts)
 
-    // Generate IO invalidation commands to initialize outputs as unused
-    val invalidateCommands = getModulePorts map {port => DefInvalid(UnlocatableSourceInfo, port.ref)}
-
+    // Generate IO invalidation commands to initialize outputs as unused,
+    //  unless the client wants explicit control over their generation.
+    val invalidateCommands = {
+      if (!compileOptions.explicitInvalidate) {
+        getModulePorts map { port => DefInvalid(UnlocatableSourceInfo, port.ref) }
+      } else {
+        Seq()
```

Comment:
I don't understand your comment here. This is the fundamental feature of the InvalidateAPI: if enabled, user code must explicitly generate "is invalid" using DontCare connections, otherwise, we generate blanket "is invalid" - the current (pre-InvalidateAPI) behavior. This is where we deal with Module IO ports.

Reply:
Ok, so this is kind of a compatibility switch for previous code that didn't have this in mind.

Reply:
Correct.